### ISSUE TITLE
[vim-slime + tmux + vim + ipython] Issue when copy multiple lines; newline character appears as ^M

### ISSUE THREAD

**Issue by pvmilk** [jpalardy/vim-slime#245]
I have issue when using vim-slime to send python code to ipython.
It only happens when I send multiple line of codes (more than 1 line). 
The newline character simply appear as ^M in the ipython screen. 
It seems that the format of code was correct, just that the newline is converted to DOS/Windows newline character. 

![image](https://user-images.githubusercontent.com/1500082/82925432-ab012800-9fb8-11ea-80d1-e043e768ba72.png)

In addition
- It doesn't not always happen. Usually, it works fine initially. After sending some multiple line of codes for several times, it appears. 
- Also, my ipython interpreter is in the docker environment. Usually, after I restart the docker container, it works fine again. 

I knew that I had a quite complicated setup, so I am trying to debug by myself also. 
I think newline character was converted to ^M at some buffer, but I don't which one. 
Can you help suggesting where should I look?

ENV:
- [host] vim-slime commit a81139239d6447b1c021e4873e6a14b47f32b65a
- [host] VIM - Vi IMproved 8.2 (2019 Dec 12, compiled May  2 2020 12:57:56)
- [host] tmux 2.7
- [docker] Python 3.6.9 (default, Apr 18 2020, 01:56:04)
- [docker] IPython 7.13.0

**Comment by jpalardy** [comment]
Hi @pvmilk,

There's surprisingly little relevant code in vim-slime 😄 

- [tmux send code](https://github.com/jpalardy/vim-slime/blob/master/autoload/slime.vim#L64)
- [python ftplugin code](https://github.com/jpalardy/vim-slime/blob/master/ftplugin/python/slime.vim)

As for the best way to debug, a generous use of `:echom` and `:messages` ([details](https://learnvimscriptthehardway.stevelosh.com/chapters/01.html))

Let me know what you find and if I can help.

**Comment by jessebett** [comment]
I'm also having this issue. I have also set `:let g:slime_python_ipython=1`.

**Comment by pvmilk** [comment]
@jessebett Finally after a long search, I found that you can use [Ctrl + d] to exit that `%cpaste` session. You might need to press it a couple of time and wait a few seconds, but it definitely better than the need to kill the process (which I also used that for a while also). 

I am still searching for where the conversion occurred. Probably will take a while as I am not so familiar with vimscript and tmux buffer. Will report back.

**Comment by jessebett** [comment]
Ah! multiple `ctrl+d`! Thank you now I can avoid nuking my interactive sessions whenever this happens.

**Comment by jpalardy** [comment]
I was looking over the tmux send code and the tmux man page.

```
     paste-buffer [-dpr] [-b buffer-name] [-s separator] [-t target-pane]
           (alias: pasteb)
         Insert the contents of a paste buffer into the specified pane.
         If not specified, paste into the current one.  With -d, also
         delete the paste buffer.  When output, any linefeed (LF) charac-
         ters in the paste buffer are replaced with a separator, by
         default carriage return (CR).  A custom separator may be speci-
         fied using the -s flag.  The -r flag means to do no replacement
         (equivalent to a separator of LF).  If -p is specified, paste
         bracket control codes are inserted around the buffer if the
         application has requested bracketed paste mode.
```

could you try the `-r` and `-p` flags (all combinations) and tell me if that improves things or not?

**Comment by pvmilk** [comment]
Thank you very much for looking into this. 

Do you mean that I should try to include the flag in the vim-slime source code directly?
In either case, this should take a while. I have not find a way to reproduce the bugs yet.

**Comment by jpalardy** [comment]
@pvmilk,

Yes, modify the vim-slime in your git clone and see what happens 😃

**Comment by pvmilk** [comment]
@jessebett After you use the trick (crlt + d multiple times), and would like the ipython + vim-slime to work again in the next run, you can type
```
$ stty sane
```
REF: https://askubuntu.com/a/452576

@jpalardy There is a possibility that there might be nothing to do with vim-slime. You may close the issue, but I will get to you once I have time to debug what you mentioned above.

**Comment by jpalardy** [comment]
@pvmilk, no problem.

Let's keep this one open — it will auto-close if it's around long enough.

Feel free to update this ticket, or open a new one 😄
